torch/nn/modules/distance.py

```diff
@@ -1,6 +1,39 @@
-
+import torch
+from .module import Module
 # TODO: Cosine
-# TODO: CosineDistance - make sure lua's CosineDistance isn't actually cosine similarity
+
+
+class CosineDistance(Module):
+    """
+    Computes cosine similarity between two vectors x1,x2:
+
+          x1*x2
+        ---------
+        |x1|*|x2|
+
+    """
+
+    def __init__(self):
+        super(CosineDistance, self).__init__()
+        self.eps = 1e-12
+
+    def forward(self, x):
+        assert len(x) == 2, "Input needs to be two vectors"
+        x1,x2 = x
+        if x1.dim() == 1 or x2.dim() == 1:
+            x1 = x1.view(1, -1)
+            x2 = x2.view(1, -1)
+        #
+        #	 x1 * x2	 w12
+        #	--------- = -----
+        #	|x1|*|x2|	w1*w2
+        #
+        # Nominator
+        w12 = torch.sum(x1 * x2, 1)
+        w1 = 1. / (torch.norm(x1, 2, 1) + self.eps)
+        w2 = 1. / (torch.norm(x2, 2, 1) + self.eps)
```

Comment:
I think it's better to add eps and divide only once - compute `w1` and `w2` as norms and make the final expression be `w12 / ((w1 * w2) + self.eps).